homeassistant/components/local_todo/todo.py

```diff
@@ -0,0 +1,162 @@
+    try:
+        return Todo.parse_obj(dataclasses.asdict(item, dict_factory=_todo_dict_factory))
+    except ValidationError as err:
+        _LOGGER.debug("Error parsing todo input fields: %s (%s)", item, str(err))
+        raise HomeAssistantError("Error parsing todo input fields") from err
+
+
+class LocalTodoListEntity(TodoListEntity):
+    """A To-do List representation of the Shopping List."""
+
+    _attr_has_entity_name = True
+    _attr_supported_features = (
+        TodoListEntityFeature.CREATE_TODO_ITEM
+        | TodoListEntityFeature.DELETE_TODO_ITEM
+        | TodoListEntityFeature.UPDATE_TODO_ITEM
+        | TodoListEntityFeature.MOVE_TODO_ITEM
+    )
+    _attr_should_poll = False
+
+    def __init__(
+        self,
+        store: LocalTodoListStore,
+        calendar: Calendar,
+        name: str,
+        unique_id: str,
+    ) -> None:
+        """Initialize LocalTodoListEntity."""
+        self._store = store
+        self._calendar = calendar
+        self._attr_name = name.capitalize()
+        self._attr_unique_id = unique_id
+
+    async def async_update(self) -> None:
+        """Update entity state based on the local To-do items."""
+        self._attr_todo_items = [
+            TodoItem(
+                uid=item.uid,
+                summary=item.summary or "",
+                status=ICS_TODO_STATUS_MAP.get(
+                    item.status or TodoStatus.NEEDS_ACTION, TodoItemStatus.NEEDS_ACTION
+                ),
+            )
+            for item in self._calendar.todos
+        ]
+
+    async def async_create_todo_item(self, item: TodoItem) -> None:
+        """Add an item to the To-do list."""
+        todo = _convert_item(item)
+        TodoStore(self._calendar).add(todo)
+        await self._async_save()
+        await self.async_update_ha_state(force_refresh=True)
+
+    async def async_update_todo_item(self, item: TodoItem) -> None:
+        """Update an item to the To-do list."""
+        todo = _convert_item(item)
+        TodoStore(self._calendar).edit(todo.uid, todo)
+        await self._async_save()
+        await self.async_update_ha_state(force_refresh=True)
+
+    async def async_delete_todo_items(self, uids: list[str]) -> None:
+        """Add an item to the To-do list."""
+        store = TodoStore(self._calendar)
+        for uid in uids:
+            store.delete(uid)
+        await self._async_save()
+        await self.async_update_ha_state(force_refresh=True)
+
+    async def async_move_todo_item(self, uid: str, pos: int) -> None:
+        """Re-order an item to the To-do list."""
+        todos = self._calendar.todos
+        found_item: Todo | None = None
+        for idx, itm in enumerate(todos):
+            if itm.uid == uid:
+                found_item = itm
+                todos.pop(idx)
+                break
+        if found_item is None:
+            raise HomeAssistantError(
+                f"Item '{uid}' not found in todo list {self.entity_id}"
```

Comment:
Is this an api error or a user error?

Reply:
Let's keep it like this for now. If it would be a service the current convention is to raise ValueError for bad user input.